### PR TITLE
POSIX sockets: implicit bind for connect()

### PR DESCRIPTION
**Pull request by OlegHahm**
According to http://pubs.opengroup.org/onlinepubs/009695399/functions/connect.html for a "socket [that] has not already been bound to a local address, connect() shall bind it to an address which, unless the socket's address family is AF_UNIX, is an unused local address."

~~Depends on #4385.~~

### CODE REVIEW

**Comment by cgundogan** [comment]
needs a rebase

**Comment by OlegHahm** [comment]
rebased

File: sys/posix/sockets/posix_sockets.c [hunk 1]
```diff
@@ -467,6 +463,30 @@ int connect(int socket, const struct sockaddr *address, socklen_t address_len)
     switch (s->type) {
 #ifdef MODULE_CONN_TCP
         case SOCK_STREAM:
+            /* "If the socket has not already been bound to a local address,
```
Comment (cgundogan) [inline sys/posix/sockets/posix_sockets.c]: this looks a lot like that what you introduced for `SOCK_DGRAM`.
Could you refactor this out of the switch? It seems to be socket type agnostic.

Comment (OlegHahm) [inline sys/posix/sockets/posix_sockets.c]: addressed

**Comment by cgundogan** [comment]
please squash and merge after travis agrees

**Comment by OlegHahm** [comment]
Squashed

**Comment by OlegHahm** [comment]
Anyone understanding why Travis is failing?

**Comment by cgundogan** [comment]
> This job ran on our new platform for Legacy Precise builds. Please read our blog post for more information.

Seems like another git version?
If I call `git log -1 --pretty=format:%H riot/master` with my local git then I get the same error

**Comment by cgundogan** [comment]
https://blog.travis-ci.com/2015-11-27-moving-to-a-more-elastic-future

**Comment by OlegHahm** [comment]
> Seems like another git version?
If I call git log -1 --pretty=format:%H riot/master with my local git then I get the same error.

Me too, but that's quite normal, since I don't have a remote called "riot" configured. But Travis should.

**Comment by cgundogan** [comment]
Yea, you are right. I did the same steps that travis does locally and it works

**Comment by OlegHahm** [comment]
Requires https://github.com/RIOT-OS/RIOT/pull/4423

**Comment by cgundogan** [comment]
travis disagrees

**Comment by OlegHahm** [comment]
> travis disagrees

?

**Comment by cgundogan** [comment]
```
error: '_implicit_bind' defined but not used [-Werror=unused-function]
 static int _implicit_bind(socket_t *s, void *addr)
```

**Comment by cgundogan** [comment]
I see the connection to #4423 now

**Comment by OlegHahm** [comment]
That exactly what I said: https://github.com/RIOT-OS/RIOT/pull/4386#issuecomment-162530056

**Pull request merged**